source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,154 @@
+
+TODO: Description of the UI actions once they are implemented.
+
+## Testing
+
+There are two areas to test: TPM presence and data persistence.
+
+TPM devices should be present in VMs running in clusters of versions >= 4.4 (TODO: re-check the version) when the corresponding cluster option is checked.  A TPM device is present as ``<tpm>`` device in the domain XML and as a corresponding device in the guest OS (``/dev/tpm0`` on Linux systems).
+
+Newly created VMs, or VMs that didn't have a TPM device previously, get uninitialized TPM data, there is nothing to test about it.  Once TPM data is modified by the guest OS, the modified data must be present after different operations: VM stop & start, reboot, migration, export/import, cloning, start from snapshot, etc.  TPM data updates followed by common failure scenarios (disconnected host, crashed VM, etc.) can be tested in situations where it's reasonable to assume the data updates should be available.  The data can be checked in the guest OS (most reliably) or alternatively comparing swtpm data with expected contents.
+
+It should be also tested that VMs with TPM data created on hosts running older systems are unchanged when the VM is started on a host running the newest system.  This must be tested in the guest OS.
+
+## Changes to Engine and Vdsm
+
+There are no special changes needed in order to add a TPM device as such, it's just another kind of device added to Engine.
+
+### Data persistence
+
+A very different thing is ensuring TPM data persistence.  TPM data is managed using an external tool, swtpm.  Once the VM domain XML is deleted on the host, the TPM data is deleted as well by libvirt and it must be restored before the VM is run again, whether on this or another host (live host migrations are handled transparently and don't require any special actions on the oVirt side).
+
+There are several options how to provide TPM data on the VM start:
+
+- As a parameter to a VM.create API call.
+- In metadata section of the domain XML.
+- On a shared storage.
+
+The hard part is how to store updated TPM data from the host reliably, making sure that:
+
+- Data updates are not lost.
+- The updated data is consistent, i.e. it can be read by swtpm correctly on the next VM start.
+
+We cannot handle all erroneous situations (e.g. faulty hardware corrupting data) but a reasonable level of resistance against common failures should be granted.  Ideally, the following situations should be handled:
+
+- When a VM stops for any reason on a connected host, we store the latest state of the data.
+- When a host or a storage get disconnected, we store the latest known state of the data.
+- When a VM is restarted without Engine assistance, it uses the final data from the last run.
```

Comment:
We cannot support this. Engine assistance is always required to run vms, for example for attaching storage to host.

Reply:
If the storage is already attached (on the same host where the VM was running last time) then it may be possible. It's a completely different topic though, all we need here now is not to close the door to the functionality with TPM handling.